From f73bc9789be7b2c0131666c503054161ef34e744 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 1 Oct 2024 23:14:36 +0200
Subject: [PATCH 1269/1695] libutils: add bisect helper functions

Implement a generic array bisecting helper function, bisect_equal().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 9ca89aa20112e003f48496e32798491ad143d32b libutils: add bisect helper functions]
Change-Id: Ic0915b994fbed45fc3f3d7efd2b50991676166a1
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/408382
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 lib/libutils/isoc/bisect.c         | 42 ++++++++++++++++++++++++++++++
 lib/libutils/isoc/include/bisect.h | 36 +++++++++++++++++++++++++
 lib/libutils/isoc/sub.mk           |  1 +
 3 files changed, 79 insertions(+)
 create mode 100644 lib/libutils/isoc/bisect.c
 create mode 100644 lib/libutils/isoc/include/bisect.h

diff --git a/lib/libutils/isoc/bisect.c b/lib/libutils/isoc/bisect.c
new file mode 100644
index 000000000..960edf1ab
--- /dev/null
+++ b/lib/libutils/isoc/bisect.c
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+
+#include <bisect.h>
+
+void *bisect_equal(const void *array, size_t n, size_t cell_size,
+		   const void *target, int (*cmp)(const void *, const void *))
+{
+	const uint8_t *array_base = array;
+	size_t high_index = 0;
+	size_t low_index = 0;
+
+	if (!array || !n)
+		return NULL;
+
+	high_index = n - 1;
+
+	if (cmp(array_base + high_index * cell_size, target) < 0)
+		return NULL;
+	if (cmp(array_base + low_index * cell_size, target) > 0)
+		return NULL;
+
+	while (high_index > low_index) {
+		size_t index = (high_index + low_index) / 2;
+		int diff = cmp(array_base + index * cell_size, target);
+
+		if (!diff)
+			return (void *)(array_base + index * cell_size);
+
+		if (diff > 0)
+			high_index = index - 1;
+		else
+			low_index = index + 1;
+	}
+
+	if (cmp(array_base + low_index * cell_size, target) == 0)
+		return (void *)(array_base + low_index * cell_size);
+
+	return NULL;
+}
diff --git a/lib/libutils/isoc/include/bisect.h b/lib/libutils/isoc/include/bisect.h
new file mode 100644
index 000000000..dcd08a737
--- /dev/null
+++ b/lib/libutils/isoc/include/bisect.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+#ifndef __BISECT_H
+#define __BISECT_H
+
+#include <stddef.h>
+#include <stdlib.h>
+
+/*
+ * Sort array in increasing order according to @cmp rule
+ * @array: Array to sort for bisect support
+ * @n: Number of cells in @array
+ * @cell_size: Byte size of a single cell of @array
+ * @cmp: Trilean comparision helper function applicable to @array
+ */
+static inline void bisect_sort(void *array, size_t n, size_t cell_size,
+			       int (*cmp)(const void *, const void *))
+{
+	qsort(array, n, cell_size, cmp);
+}
+
+/*
+ * Bisect into array to find the cell that matches an entry
+ * @array: Sorted array (according the @cmp) to bisect in
+ * @n: Number of cells in @array
+ * @cell_size: Byte size of a single cell of @array
+ * @target: Entry for which we find an equal entry (with @cmp) in sorted @array
+ * @cmp: Trilean comparision helper function applicable to @array
+ * @return: Pointer to the cell in @array matching @target, NULL if none found
+ */
+void *bisect_equal(const void *array, size_t n, size_t cell_size,
+		   const void *target, int (*cmp)(const void *, const void *));
+
+#endif /*__BISECT_H*/
diff --git a/lib/libutils/isoc/sub.mk b/lib/libutils/isoc/sub.mk
index ef1ca5da8..45c14cbc4 100644
--- a/lib/libutils/isoc/sub.mk
+++ b/lib/libutils/isoc/sub.mk
@@ -27,6 +27,7 @@ srcs-y += islower.c
 srcs-y += isprint.c
 srcs-y += ispunct.c
 srcs-y += toupper.c
+srcs-y += bisect.c
 
 ifneq (,$(filter ta_%,$(sm)))
 srcs-y += fp.c
-- 
2.47.2

